File: lib/packet.cpp
#include "packet.hpp"

#include <utility>

/**
 * Copy Byte
 * Copies src data into the buffer
 *
 */
void copy_bytes(uint8_t* buf, const uint8_t* src, size_t n) {
    for (size_t i = 0; i < n; i++) {
        buf[i] = src[i];
    }
}

void serialize_packet(Packet p, uint8_t buf[PACKET_SIZE]) {
    uint32_t seq_no = htonl(p.seq_no);
    copy_bytes(buf + 0, (uint8_t*)&seq_no, 4);

    union _u128_as_2u64 {
        __uint128_t v;
        uint32_t u[4];
    } chksm;
    chksm.v = p.checksum;
    chksm.u[0] = htonl(chksm.u[0]);
    chksm.u[1] = htonl(chksm.u[1]);
    chksm.u[2] = htonl(chksm.u[2]);
    chksm.u[3] = htonl(chksm.u[3]);
    swap(chksm.u[0], chksm.u[3]);
    swap(chksm.u[1], chksm.u[2]);
    copy_bytes(buf + 4, (uint8_t*)&chksm.v, 16);
    uint8_t anl = 0;
    if (p.ack) {
        anl |= (1 << 7);
    }
    if (p.nak) {
        anl |= (1 << 6);
    }
    anl |= p.length;
    copy_bytes(buf + 20, &anl, 1);
    copy_bytes(buf + 21, p.data, 43);
}

Packet deserialize_packet(const uint8_t data[PACKET_SIZE]) {
    Packet p;
    p.seq_no = ntohl(*(uint32_t*)data);
    union _u128_as_2u64 {
        __uint128_t v;
        uint32_t u[4];
    } chksm;
    chksm.u[0] = ntohl(*((uint32_t*)data + 1));
    chksm.u[1] = ntohl(*((uint32_t*)data + 2));
    chksm.u[2] = ntohl(*((uint32_t*)data + 3));
    chksm.u[3] = ntohl(*((uint32_t*)data + 4));
    p.checksum = chksm.v;
    uint8_t anl = data[20];
    p.ack = anl & (1 << 7);
    p.nak = anl & (1 << 6);
    p.length = anl & (0b00111111);
    copy_bytes(p.data, data + 21, p.length);
    return p;
}

File: src/packet.py
from dataclasses import dataclass

from hashlib import md5

MAX_PACKET_SIZE: int = 64
DATA_SIZE: int = 43
CONTINUATION_PREFIX: bytes = b"\xFE\xFD"
SEQ_LIM = 2**32
FINISHER_DATA = b"TEKCAP TSAL"


def hexdigest_to_bytes(digest: str) -> bytes:
    r = b""
    assert len(digest) % 2 == 0
    for i in range(len(digest) // 2):
        a = digest[2 * i]
        b = digest[2 * i + 1]
        a = int(a, 16)
        b = int(b, 16)
        c = (a << 4) | b
        r += c.to_bytes(1, 'big', signed=False)
    return r


@dataclass
class Packet:
    seq_no: int
    ack: bool
    nak: bool
    data: bytes

    def into_buf(self) -> bytes:
        buf = b""
        buf += self.seq_no.to_bytes(4, 'big', signed=False)
        # print(self.data)
        chksm = md5(self.data).hexdigest()
        buf += hexdigest_to_bytes(chksm)
        anl = 0
        if self.ack: anl |= 0b10_00_00_00
        if self.nak: anl |= 0b01_00_00_00
        anl |= len(self.data)
        buf += anl.to_bytes(1, 'big', signed=False)
        buf += self.data
        buf += b'\0' * (64 - len(buf))
        return buf

    @staticmethod
    def from_buf(buf: bytes):
        assert len(buf) == 64
        seq_no = int.from_bytes(buf[:4], 'big', signed=False)
        chksm = buf[4:20]
        anl = buf[20]
        ack = (anl & 0b10_00_00_00) != 0
        nak = (anl & 0b01_00_00_00) != 0
        length = anl & 0b00_11_11_11
        data = buf[21:21 + length]
        assert chksm == hexdigest_to_bytes(md5(data).hexdigest())
        return Packet(seq_no, ack, nak, data)

    @staticmethod
    def finisher(seq_no: int):
        return Packet(seq_no, True, True, FINISHER_DATA)

    @staticmethod
    def acknowledgment(seq_no: int):
        return Packet(seq_no, True, False, b"")
